packages/hash/api/src/model/knowledge/user.model.ts

```diff
@@ -0,0 +1,163 @@
+  title: "User",
+  properties: [
+    {
+      baseUri: shortnameBaseUri,
+      versionedUri: shortnamePropertyType.$id,
+    },
+    {
+      baseUri: emailBaseUri,
+      versionedUri: emailPropertyType.$id,
+      required: true,
+      array: { minItems: 1 },
+    },
+  ],
+});
+
+const userEntityTypeVersionedUri = userEntityType.$id;
+
+/**
+ * @class {@link UserModel}
+ */
+export default class extends EntityModel {
+  /**
+   * Get a workspace user entity by their shortname.
+   *
+   * @param params.shortname - the shortname of the user
+   */
+  static async getUserByShortname(
+    graphApi: GraphApi,
+    params: { shortname: string },
+  ): Promise<UserModel | null> {
+    /** @todo: use upcoming Graph API method to filter entities in the datastore */
+    const allEntities = await EntityModel.getAllLatest(graphApi, {
+      accountId: workspaceAccountId,
+    });
+
+    const matchingEntity = allEntities
+      .filter(
+        ({ entityTypeModel }) =>
+          entityTypeModel.schema.$id === userEntityTypeVersionedUri,
+      )
+      .find(
+        ({ properties }) =>
+          (properties as any)[shortnameBaseUri] === params.shortname,
+      );
+
+    return matchingEntity ? new UserModel(matchingEntity) : null;
+  }
+
+  static async getUserEntityType(graphApi: GraphApi): Promise<EntityTypeModel> {
+    return await EntityTypeModel.get(graphApi, {
+      versionedUri: userEntityTypeVersionedUri,
+    });
+  }
+
+  /**
+   * Create a workspace user entity.
+   *
+   * @param params.emails - the emails of the user
+   * @param params.shortname - the shortname of the user
+   */
+  static async createUser(
+    graphApi: GraphApi,
+    params: UserModelCreateParams,
+  ): Promise<UserModel> {
+    const { emails, shortname } = params;
+
+    // if setting a shortname, ensure it's unique across all workspace users
+    if (shortname) {
+      const existingUserWithShortname = await UserModel.getUserByShortname(
+        graphApi,
+        { shortname },
+      );
+
+      if (existingUserWithShortname) {
+        throw new Error(
+          `A user entity with shortname "${shortname}" already exists.`,
+        );
+      }
+
+      /** @todo: also ensure shortname is unique amongst orgs */
+    }
+
+    const properties: object = {
+      [emailBaseUri]: emails,
+      [shortnameBaseUri]: shortname,
+    };
+
+    const entityTypeModel = await UserModel.getUserEntityType(graphApi);
+
+    const accountId = workspaceAccountId;
+
+    const { entityId, version } = await EntityModel.create(graphApi, {
+      accountId,
+      properties,
+      entityTypeModel,
+    });
+
+    return new UserModel({
+      accountId,
+      entityId,
+      version,
+      entityTypeModel,
+      properties,
+    });
+  }
+
+  getEmails(): string[] {
+    return (this.properties as any)[emailBaseUri];
+  }
+
+  getShortname(): string {
+    return (this.properties as any)[shortnameBaseUri];
+  }
```

Comment:
I'm really sad about not being able to type the `properties` object in this class, if anyone has suggestions/ideas how to improve this please let me know.

cc @hashintel/backend

Reply:
Other than using other type definitions that we create from TS outside the SDK, I'm not sure what else we can do. Perhaps this gets better with the "builders" we're planning to have?

Reply:
I think it will get better with the builders, the frustration is acknowledged, lets see how it develops with the upcoming pieces :)